### PR TITLE
Date

### PR DESCRIPTION
**Pull request by nastydrew**

### CODE REVIEW

File: Date/Date.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#include <iostream>
+#include "Date.h"
+
+using namespace std;
+
+void Date::validate(int day, int month, int year) {
+    int daysInMonth[13] = {1, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+    if ( day > daysInMonth[month] || day < daysInMonth[0]) {
```
Comment (igordotsenko) [inline Date/Date.cpp]: `day < daysInMonth[0]` in this case is not very obvious.`day < 1` would be better

File: Date/Date.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        throw InvalidDateException("Invalid Day!");
+    }
+    if ( month < 1 || month > 12 ) {
+        throw InvalidDateException("Invalid Month!");
+    }
+    if ( year < 0 ) {
```
Comment (igordotsenko) [inline Date/Date.cpp]: Why? ) Did you hear anything about B.C.? )

File: Date/Date.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        throw InvalidDateException("Invalid Year!");
+    }
+    if ( month == 2 || month == 4 || month == 6 || month == 9 || month == 11 ) {
+        if ( day == 31) {
+            throw InvalidDateException("Invalid Month!");
+        }
+    }
```
Comment (igordotsenko) [inline Date/Date.cpp]: Come on =(
`if ( day > daysInMonth[month] || day < daysInMonth[0]) {` is not enough for you?

File: Date/Date.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    if ( day == 30 && month == 2 ) {
+        throw InvalidDateException("Invalid Month!");
+    }
```
Comment (igordotsenko) [inline Date/Date.cpp:19]: =((((

File: Date/Date.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if ( year % 4 != 0 ) {
+        if  ( day == 29 && month == 2 ) {
+            throw InvalidDateException("Error, February has 28 days!");
+        }
+    }
+}
```
Comment (igordotsenko) [inline Date/Date.cpp:25]: This formula is not full. Google better

File: Date/Date.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
+#include <iostream>
+#include "Date.h"
+
+using namespace std;
+
+void Date::validate(int day, int month, int year) {
+<<<<<<< HEAD
```
Comment (igordotsenko) [inline Date/Date.cpp]: Ohohoho )

File: Date/Date.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    if ( day > daysInMonth[month] || day < 1 ) {
+        throw InvalidDateException("Invalid Day!");
+    }
+    if ( month < 1 || month > 12 ) {
+        throw InvalidDateException("Invalid Month!");
+    }
+    if ( year == 0 ) {
+        throw InvalidDateException("Invalid Year!");
+    }
+    if ( month == 4 || month == 6 || month == 9 || month == 11 ) {
```
Comment (igordotsenko) [inline Date/Date.cpp]: Why?

File: Date/Date.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+using namespace std;
+
+void Date::validate(int day, int month, int year) {
+    int daysInMonth[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+    if ( day > daysInMonth[month] || day < 1 ) {
+        throw InvalidDateException("Invalid Day!");
+    }
+    if ( month < 1 || month > 12 ) {
```
Comment (igordotsenko) [inline Date/Date.cpp]: You should do this validation first. In other case ArrayOutOfBoundException can be thrown here: `if ( day > daysInMonth[month] || day < 1 ) {`

File: Date/Date.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
+#include <iostream>
+#include "Date.h"
+
+using namespace std;
+
+void Date::validate(int day, int month, int year) {
+    int daysInMonth[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+    if ( day > daysInMonth[month] || day < 1 ) {
```
Comment (igordotsenko) [inline Date/Date.cpp]: What if this year February has 29 days?

File: Date/Date.cpp (hunk 7 above)
Comment (nastydrew) [inline Date/Date.cpp]: What do you mean?

Comment (igordotsenko) [inline Date/Date.cpp]: `if ( day > daysInMonth[month] || day < 1 ) {`

Is that not enough?

File: Date/Date.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        throw InvalidDateException("Invalid Month!");
+    }
+}
+
+Date::Date(int day, int month, int year) {
+    this->day = day;
+    this->month = month;
+    this->year = year;
+
+    validate(day, month, year);
```
Comment (igordotsenko) [inline Date/Date.cpp]: My logic tells me that validation should be before values assigning to object fields

File: Date/Date.cpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+void Date::validate(int day, int month, int year) {
+    int daysInMonth[MONTH_AMOUNT] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+    int indexMonth = month - 1;
+
+    if ( year % 400 == 0 || year % 4 == 0 && year % 100 != 0 ) {
+        daysInMonth[FEBRUARY] += 1;
+    }
+
+    if ( day > daysInMonth[indexMonth] || day < 1 ) {
```
Comment (igordotsenko) [inline Date/Date.cpp:21]: What if user passes month = 100500?
Oops...
You should test it better. On real job your team lead will be very sad if you will fix the same bug 1488 times

File: Vector/main.cpp [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    cout << "Length: " << a.len(b) << endl;
+    cout << "Summary:" << a + b << endl;
+    cout << "Different: " << a - b << endl;
+
+
+
+    return 0;
+}
```
Comment (igordotsenko) [inline Vector/main.cpp]: Some new files for Vector? )

File: List/List.cpp [hunk 13]
```diff
@@ -176,4 +176,4 @@ ostream& operator<<(std::ostream& out, const List<T>& list) {
     out << list[last];
     
     return out;
-}
+}
```
Comment (igordotsenko) [inline List/List.cpp]: Come on

**Pull request merged**